buildSrc/src/main/kotlin/org/partiql/gradle/plugin/pig/PigPlugin.kt

```diff
@@ -0,0 +1,218 @@
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package org.partiql.gradle.plugin.pig
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.options.Option
+import org.partiql.pig.main
+import java.io.File
+import javax.inject.Inject
+
+/**
+ * Imported from https://github.com/partiql/partiql-ir-generator/tree/main/pig-gradle-plugin
+ *
+ * TODO REMOVE ONCE `pig-gradle-plugin` is published
+ */
+abstract class PigPlugin : Plugin<Project> {
+
+    override fun apply(target: Project) {
+        // Ensure `sourceSets` extension exists
+        target.pluginManager.apply(JavaPlugin::class.java)
+
+        // Adds pig source set extension to all source sets
+        target.sourceSets().forEach { sourceSet ->
+            val name = sourceSet.name
+            val sds = target.objects.sourceDirectorySet(name, "$name PIG source")
+            sds.srcDir("src/$name/pig")
+            sds.include("**/*.ion")
+            sourceSet.extensions.add("pig", sds)
+        }
+
+        // Extensions for pig compiler arguments
+        val ext = target.extensions.create("pig", PigExtension::class.java)
+
+        // Create tasks after source sets have been evaluated
+        target.afterEvaluate {
+            target.sourceSets().forEach { sourceSet ->
+                // Pig generate all for the given source set
+                val pigAllTaskName = getPigAllTaskName(sourceSet)
+                val pigAllTask = target.tasks.create(pigAllTaskName) {
+                    this.group = "pig"
+                    this.description = "Generate all PIG sources for ${sourceSet.name} source set"
+                }
+
+                val language = ext.language
+                val namespace = ext.namespace
+                var outDir = ext.outputDir
+
+                // The `kotlin` target without an `outputDir` will use the package path as the output dir
+                if (outDir == null && language == "kotlin") {
+                    val base = "${target.buildDir}/generated-src"
+                    outDir = File("$base/${namespace.replace('.', '/')}")
+                    // Eventually this will need to be sourceSet.kotlin, but that's currently not possible
+                    // sourceSet.java.srcDir(base)
```

Comment:
Right now you can mark Kotlin sources as a "java" sourceset. At some point, Gradle won't allow that (according to forum posts I can't hunt down right now). Not pressing and we don't have control over this. I just left that note to say "Hey, I'm marking this generated Kotlin code as 'java' src because it's the only current option and works for the foreseeable future, but there's talk on the horizon of this no longer working"